src/nvim/eval.c

```diff
@@ -18095,45 +18096,147 @@ void free_tv(typval_T *varp)
   }
 }
 
-/*
- * Free the memory for a variable value and set the value to NULL or 0.
- */
+#define TYPVAL_ENCODE_ALLOW_SPECIALS false
+
+#define TYPVAL_ENCODE_CONV_NIL() \
+    do { \
+      tv->vval.v_special = kSpecialVarFalse; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_BOOL(ignored) \
+    TYPVAL_ENCODE_CONV_NIL()
+
+#define TYPVAL_ENCODE_CONV_NUMBER(ignored) \
+    do { \
+      (void) ignored; \
+      tv->vval.v_number = 0; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_UNSIGNED_NUMBER(ignored) \
+    assert(false)
+
+#define TYPVAL_ENCODE_CONV_FLOAT(ignored) \
+    do { \
+      tv->vval.v_float = 0; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_STRING(str, ignored) \
+    do { \
+      xfree(str); \
+      tv->vval.v_string = NULL; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_STR_STRING(ignored1, ignored2)
+
+#define TYPVAL_ENCODE_CONV_EXT_STRING(ignored1, ignored2, ignored3)
+
+#define TYPVAL_ENCODE_CONV_FUNC(fun) \
+    do { \
+      func_unref(fun); \
+      if (fun != empty_string) { \
+        xfree(fun); \
+      } \
+      tv->vval.v_string = NULL; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_EMPTY_LIST() \
+    do { \
+      list_unref(tv->vval.v_list); \
+      tv->vval.v_list = NULL; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_EMPTY_DICT() \
+    do { \
+      dict_unref(tv->vval.v_dict); \
+      tv->vval.v_dict = NULL; \
+      tv->v_lock = VAR_UNLOCKED; \
+    } while (0)
+
+#define TYPVAL_ENCODE_CONV_LIST_START(ignored) \
+    do { \
+      if (tv->vval.v_list->lv_refcount > 1) { \
+        tv->vval.v_list->lv_refcount--; \
+        tv->vval.v_list = NULL; \
```

Comment:
This macro is called before we push a new element on the stack in `nothing_convert_one_value()`.
When  constructing the pushed element, `tv->vval.v_list->lv_first` is read after `tv->vval.v_list` was set to `NULL`.